Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h

```diff
@@ -631,11 +637,18 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
         if (handleConstantBinaryBitwiseOp(node))
             break;
         
-        if (node->op() == ArithBitAnd
-            && (isBoolInt32Speculation(forNode(node->child1()).m_type) ||
-                isBoolInt32Speculation(forNode(node->child2()).m_type))) {
-            setNonCellTypeForNode(node, SpecBoolInt32);
-            break;
+        if (node->op() == ArithBitAnd) {
+            ASSERT(node->child1().useKind() != UntypedUse);
+            if (node->child2()->isInt32Constant() && !node->child2()->asInt32()) {
+                m_state.setShouldTryConstantFolding(true);
+                forNode(node) = forNode(node->child1());
```

Comment:
`x & 0` is always 0, right? So, should it be this?
```
forNode(node) = forNode(node->child2());
```

Reply:
Changed.